README.md

```diff
@@ -15,58 +15,146 @@ The library ships with a number of features:
+the following schema:
 
 ```python
-ncbi_tax = Taxonomy.from_ncbi('./nodes.dmp', './names.dmp')
-assert tax.name('562') == 'Escherichia coli'
-assert tax.rank('562') == 'species'
+class TaxonomyNode:
+    id: str
+    name: str
+    parent: Optional[str]
+    rank: str
 ```
 
-Note that Taxonomy IDs in NCBI format are integers, but they're converted to strings on import. We find working with "string taxonomy IDs" greatly simplifies interoperation between different taxonomy systems.
+Note that tax_id in parameters passed in functions described below are string but for example in the case of NCBI need
+to be essentially quoting integers: `562 -> "562"`. In that case, passing something that can't be converted to a number
+will raise an exception even if the documentation below does not mention it.
+
+#### `tax.root -> TaxonomyNode`
+Points to the root of the taxonomy
+
+#### `tax.parent(tax_id: str, /, at_rank: str) -> (Optional[TaxonomyNode], Optional[float])`
+Return the immediate parent TaxonomyNode of the node id provided as well as the distance to it as a `(TaxonomyNode, float)` tuple.
+
+If `at_rank` is provided, scan all the nodes in the node's lineage and return
+the parent id at that rank.
+
+Examples:
+
+```py
+parent, distance = tax.parent("612")
+parent, _ = tax.parent("612", at_rank="species")
+parent, distance = tax.parent("612")
+# Both variables will be `None` if we can't find the parent
+parent, distance = tax.parent("unknown")
+```
+
+#### `tax.find_by(tax_id: str) -> Optional[TaxonomyNode]`
+
+Returns the node at that id. Returns `None` if not found.
+You can also use indexing to accomplish that: `tax["some_id"]` but this will raise an exception if the node
+is not found.
+
+#### `tax.find_by_name(name: str) -> Optional[TaxonomyNode]`
+
+Returns the node with that name. Returns `None` if not found.
+In NCBI, it only accounts for *scientific names* and not synonyms.
+
+#### `tax.children(tax_id: str) -> List[TaxonomyNode]`
+
+Returns all nodes below the given tax id.
+
+#### `tax.lineage(tax_id: str) -> List[TaxonomyNode]`
+
+Returns all nodes above the given tax id, including itself.
+
+#### `tax.lca(id1: str, id2: str) -> Optional[TaxonomyNode]`
+
+Returns the [lowest common ancestor](https://en.wikipedia.org/wiki/Lowest_common_ancestor) for the 2 given nodes.
+
+#### `tax.prune(keep: List[str], remove: List[str])-> Taxonomy`
+
+Return a copy of the taxonomy containing:
+
+- only the nodes in `keep` and their parents if provided
+- all of the nodes except those in remove and their children if provided
+
+#### `tax.remove(tax_id: str)`
+
+Remove the node from the tree, re-attaching parents as needed: only a single node is removed.
+
+#### `tax.add(parent_tax_id: str, new_tax_id: str)`
+
+Add a new node to the tree at the parent provided.
+
+#### `edit_node(tax_id: str, /, name: str, rank: str, parent_id: str, parent_dist: float)`
```

Comment:
Do we need to do this vs. __setattr__? Also, what does it mean to edit `parent_dist`? I'd expect that to be computed from the topology.

Reply:
I don't think setattr would work here? We need to update internal Rust datastructures but we'd need to use setattr on an indexed value to make it work, eg `setattr(tax["512"], "name", "E. Coli")` which means the TaxonomyNode needs to hold a reference to the Taxonomy, which is not allowed in python class via pyo3

Reply:
Sorry I really meant `__setitem__`. Apologies for the confusion. 

I.e., can/should we have an API like:

`tax[tax_id]["name"] = "E. coli"`

Maybe not if it's not a dict-of-dicts underneath.

Reply:
That's the same issue, we don't have a reference to the taxonomy in `tax[tax_id]`, it's a plain python class